### DISCUSSION TITLE
Is there API for object keys' mapping?

### DISCUSSION THREAD

**Discussion by pyncz** [colinhacks/zod#2233] [Q&A]
Let's say, I use a 3rd party API to fetch some data. I want to parse and **transform** it with zod for further usage, including some operations with keys, e.g. mapping keys from snake_case to camelCase, or even change some confusing field names into something more convenient.

Now I would go with something like this:
```ts
const schema = z.object({
   some_key: z.string(),
}).transform(data => ({ someKey: data.some_key }))
```

Can I do that more... zod-native? I was thinking about something like this:
```ts
const schema = z.object({
   some_key: z.string().as('someKey'), // .as() or .key() or .to() or something
})
```

**Comment by pyncz** [comment]
Came up with a helper function so far:

```ts
type MappedSchema<
  T extends Record<string, any>,
  OldKey extends keyof T,
  NewKey extends string,
> = Omit<T, OldKey | NewKey> & Record<NewKey, T[OldKey]>

export const mapped = <
  T extends Record<string, any>,
  OldKey extends keyof T,
  NewKey extends string,
>(
    schema: z.ZodType<T, z.ZodTypeDef, any>,
    oldKey: OldKey,
    newKey: NewKey,
  ): z.ZodType<MappedSchema<T, OldKey, NewKey>, z.ZodTypeDef, T> => {
  return schema.transform((data) => {
    const mappedData = {} as MappedSchema<T, OldKey, NewKey>

    for (const k in data) {
      const key = k as keyof T
      if (key === oldKey) {
        mappedData[newKey as keyof MappedSchema<T, OldKey, NewKey>] = data[key]
      } else {
        mappedData[key as keyof MappedSchema<T, OldKey, NewKey>] = data[key]
      }
    }
    return mappedData
  })
}
```

```ts
const schema = mapped(mapped(z.object({
  key1: z.string(),
  key2: z.number(),
  key3: z.string(),
}), 'key2', 'key2mod'), 'key1', 'key1mod')

console.log(schema.parse({ key1: 'hey', key2: 42, key3: 'ho' }))
// > { key1mod: 'hey', key2mod: 42, key3: 'ho' }
```

**Comment by JacobWeisenburger** [comment]
Perhaps something like this would help:
https://github.com/colinhacks/zod/discussions/2240#discussioncomment-5435162

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by pyncz** [comment]
@JacobWeisenburger 

Related problem, but not exactly the same.. It's not **just** about the case of the fields.
So yeah, a regular transformation is probably the solution for now.
```ts
const schema = z.object({
   some_key: z.string(),
}).transform(data => ({ someKey: data.some_key }))
```

...Or using a helper like the one above.
I've updated its API and [shaped it into a separated package](https://github.com/pyncz/zod-key-mapper) so anyone who comes across this discussion can consider using it:
```ts
import { z } from 'zod'
import { mapped } from '@pyncz/zod-key-mapper'

const schema = mapped(z.object({
  key1: z.string(),
  key2: z.number(),
  key3: z.number(),
}), { key1: 'key1mod', key2: 'key2mod' })

console.log(schema.parse({ key1: 'hey', key2: 69, key3: 420 }))
// > { key1mod: 'hey', key2mod: 69, key3: 420 }
```
